[Feature][Connector]add AssertSink connector

lhyundeadsoul opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist

  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/incubator-seatunnel/issues).

  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.

  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.

-->

## Purpose of this pull request
 import an AssertSink plugin, this plugin is similier with ConsoleSink plugin, but will not print data in console, it will validate the data with some rule, e.g: data rows, data type, data value...

https://github.com/apache/incubator-seatunnel/issues/1912
<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

mans2singh committed:
[Bug][seatunnel-core-flink] Updated FlinkRunMode enum to get the proper help message for run modes. (#2008)

lhyundeadsoul committed:
a more understandable code,and code warning will disappear (#2005)

lhyundeadsoul committed:
first commit

tobezhou33 committed:
fix same source and sink registerplugin librarycache errpr (#2015)

Co-authored-by: zhoutao.tobeone <zhoutao.tobeone@bytedance.com>

Hisoka-X committed:
[Doc][Transform] Add uuid, udf, replace transform to doc (#2016)

* update contribute-plugin.md about plugin mapping.

* update contribute-plugin.md about plugin mapping.

* add uuid, udf, replace transform to doc

Co-authored-by: TrickyZerg <32387433+BenJFan@users.noreply.github.com>

lhyundeadsoul committed:
[Feature][Connector]add AssertSink AssertExecutor

lhyundeadsoul commented:
@ruanwenjun https://github.com/apache/incubator-seatunnel/issues/1912 has been done. Can you help me to review the code ? Thx!

lhyundeadsoul commented:
AssertSink config is at` examples/fake_to_console.conf:56` ,uncomment and it can be used by `org.apache.seatunnel.example.flink.LocalFlinkExample`

UT is here:
org.apache.seatunnel.flink.assertion.AssertExecutorTest
org.apache.seatunnel.flink.assertion.rule.AssertRuleParserTest

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-flink/pom.xml:
```diff
@@ -43,7 +43,7 @@
         <module>seatunnel-connector-flink-influxdb</module>
         <module>seatunnel-connector-flink-clickhouse</module>
-        <module>seatunnel-connector-flink-http</module>
+        <module>seatunnel-connector-flink-assert</module>
```
Please add a new module, rather than change http to assert

ruanwenjun commented on docs/sidebars.js:
```diff
@@ -128,6 +128,9 @@ const sidebars = {
         'transform/sql',
         'transform/split',
         'transform/json',
+        'transform/replace',
+        'transform/udf',
+        'transform/uuid',
```
In my knowledge, the sidebars will be auto generated(I am not sure), but you don't need to do this change in this PR.

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java:
```diff
@@ -0,0 +1,113 @@
+    @Override
+    public void outputBatch(FlinkEnvironment env, DataSet<Row> inDataSet) {
+        inDataSet.map(row -> {
+            System.out.println(row);
```
```suggestion
            
```

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public void outputBatch(FlinkEnvironment env, DataSet<Row> inDataSet) {
+        inDataSet.map(row -> {
```
It's better to use reduce here.

lhyundeadsoul replied:
Thanks for your suggestion. 
But Reduce function need to process two row data in one place, that looks a little bit strange.

ruanwenjun replied:
Why two row data? In my knowledge, reduce is like map, but it just doesn't need to return data. You don't need to get a result in inDataSet, so you return null here. You can use reduce, then you don't need to return null.

lhyundeadsoul replied:
> 
Here is reduce Defination in `org.apache.flink.api.common.functions.ReduceFunction`
`    T reduce(T value1, T value2) throws Exception;`


code will be like
`

    @Override
    public void outputBatch(FlinkEnvironment env, DataSet<Row> inDataSet) {
        inDataSet.reduce((row1, row2) -> {
            ASSERT_EXECUTOR
                    .fail(row1, assertFieldRules)
                    .ifPresent(failRule -> {
                        throw new IllegalStateException("row :" + row1 + " fail rule: " + failRule);
                    });
            ASSERT_EXECUTOR
                    .fail(row2, assertFieldRules)
                    .ifPresent(failRule -> {
                        throw new IllegalStateException("row :" + row2 + " fail rule: " + failRule);
                    });
            return null;
        });
    }
`

lhyundeadsoul replied:
code will be like
![ima](https://user-images.githubusercontent.com/5850103/174255498-d71b2362-35a1-4503-9a96-6a80dabaa7be.png)

ruanwenjun replied:
Ok, this is not important thing, I close this conversation.

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/main/java/org/apache/seatunnel/flink/assertion/sink/AssertSink.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public void outputStream(FlinkEnvironment env, DataStream<Row> dataStream) {
+        dataStream.map(row -> {
+            System.out.println(row);
```
```suggestion
        dataStream.reduce(row -> {

```

lhyundeadsoul replied:
I didn't want to `system.out` here either at first, but I found that you wouldn't know the progress of code if there is nothing output here when all data is valid. Maybe we can use LOGGER here?

There is no reduce function in dataStream API. If we want to use reduce function, we must keyBy() first.

ruanwenjun replied:
Use `System.out.println` is not reliable, it will just print the data into console. If you want to log the data here you can use LOGGER, but I think we don't need to log data here, if there is no error, means the validation is pass.

lhyundeadsoul replied:
OK, do not log here.

Please take a look this:
There is no reduce function in dataStream API. If we want to use reduce function, we must keyBy() first.

ruanwenjun commented on seatunnel-connectors/seatunnel-connectors-flink/seatunnel-connector-flink-assert/src/test/java/org/apache/seatunnel/flink/assertion/AssertExecutorTest.java:
```diff
@@ -0,0 +1,86 @@
+package org.apache.seatunnel.flink.assertion;
```
Please add license header

ruanwenjun commented on seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/core/flink/config/FlinkRunMode.java:
```diff
@@ -35,4 +35,8 @@ public enum FlinkRunMode {
     }
+
+    public String toString() {
+        return mode;
```
```suggestion
        return "FlinkRunMode{" +
            "mode='" + mode + '\'' +
            '}';
```

ruanwenjun commented on seatunnel-examples/seatunnel-flink-examples/src/main/resources/examples/fake_to_console.conf:
```diff
@@ -48,6 +48,44 @@ transform {
 sink {
   ConsoleSink {}
 
+  #AssertSink {
```
Please revert this change, and add doc for this plugin.

lhyundeadsoul replied:
Can you tell me where can add doc for this plugin?

ruanwenjun replied:
You can add log under: https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector/sink

ruanwenjun reviewed (changes requested):
Please add this plugin in `seatunnel-connectors-flink-dist/pom.xml`, otherwise the plugin will not in distribution.

lhyundeadsoul committed:
add AssertSink AssertExecutor 2

lhyundeadsoul committed:
Update seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/core/flink/config/FlinkRunMode.java

Co-authored-by: Wenjun Ruan <wenjun@apache.org>

lhyundeadsoul committed:
Merge branch 'dev' into assertplugin

lhyundeadsoul committed:
[Feature][Connector]add AssertSink AssertExecutor
1. add licenses
2. use reduce instead of map in batch output
3. revert fake_to_console.conf
4. add plugin to flink-dist

lhyundeadsoul committed:
Merge remote-tracking branch 'origin/assertplugin' into assertplugin

lhyundeadsoul committed:
[Feature][Connector]add AssertSink AssertExecutor
1. add licenses

lhyundeadsoul committed:
[Feature][Connector]add AssertSink doc

lhyundeadsoul commented:
There is so many [ImportOrder] checkstyle error.  All of them looks tricky. Is there any approach to  correct my code import order automatically? @ruanwenjun

lhyundeadsoul committed:
[Feature][Connector] obey with checkstyle

ruanwenjun commented:
> There is so many [ImportOrder] checkstyle error. All of them looks tricky. Is there any approach to correct my code import order automatically? @ruanwenjun

You need to import `tools/checkstyle/checkstyle.xml` into your IDE.

lhyundeadsoul commented:
>

lhyundeadsoul committed:
[Feature][Connector] obey with checkstyle import order

lhyundeadsoul committed:
[Feature][Connector] obey with checkstyle import order 2

ruanwenjun commented on docs/en/connector/sink/Assert.md:
```diff
@@ -0,0 +1,105 @@
+
+### field_type
+
+field type (string),  e.g. `short/int/long/string/boolean`
```
Please give the complete supported field types in the doc, don't just say 
e.g. `short/int/long/string/boolean`, if I see this, I will believe that the assert plugin only support these 5 types.

ruanwenjun commented on docs/en/connector/sink/Assert.md:
```diff
@@ -0,0 +1,105 @@
+
+The following rules are supported for now
+`
+NOT_NULL,  
+MIN,  
+MAX,  
+MIN_LENGTH,  
+MAX_LENGTH
```
Please clarify the meaning of each rule type, and if we break the rule, what will happen.

ruanwenjun commented on seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/batch/FlinkBatchExecution.java:
```diff
@@ -118,6 +118,6 @@ public Config getConfig() {
 
     private boolean whetherExecute(List<FlinkBatchSink> sinks) {
-        return sinks.stream().anyMatch(s -> !"ConsoleSink".equals(s.getPluginName()));
+        return sinks.stream().anyMatch(s -> !"ConsoleSink".equals(s.getPluginName()) && !"AssertSink".equals(s.getPluginName()));
```
In my knowledge, you cannot add this check here, this will not call execute method, so your flink job will not be submitted.

lhyundeadsoul committed:
[Feature][Connector] add more comments of Assert Sink

ruanwenjun reviewed (approved):
LGTM, you can add IT to test this plugin.

lhyundeadsoul commented:
> LGTM, you can add IT to test this plugin.

Thanks very much for reviewing. Hope to help community develop more efficiently!
